docs/design/CONFIG_API.md

```diff
@@ -0,0 +1,231 @@
+Config API Design
+===================
+
+## Motivation
+
+[Issue 51][issue_51] specifies the need to design an API for configuring Envoy Gateway. This configuration is referred
+to as the "Bootstrap Config" in the Envoy Gateway high-level [design doc][design_doc].
+
+## Goals
+
+* Define an API for configuring Envoy Gateway.
+* Specify tooling for managing the API, e.g. generate protos, CRDs, controller RBAC, etc.
+
+## Non-Goals
+
+* Implementation of the API.
+* Define the `status` subresource of the configuration API.
+
+## Proposal
+
+Utilize the [Kubebuilder][kubebuilder] framework to build and manage Envoy Gateway APIs. Kubebuilder facilitates the
+following developer workflow for building APIs:
+
+- Create one or more resource APIs as CRDs and then add fields to the resources.
+- Implement reconcile loops in controllers and watch additional resources.
+- Test by running against a cluster (self-installs CRDs and starts controllers automatically).
+- Update bootstrapped integration tests to test new fields and business logic.
+- Build and publish a container from a Dockerfile.
+
+After installing the latest Kubebuilder release, create the `BootstrapConfig` API:
+```shell
+kubebuilder create api --group envoygateway.io --version v1alpha1 --kind BootstrapConfig
+```
+
+The `v1alpha1` version and `envoygateway.io` API group get generated:
+```go
+// gateway/api/v1alpha1/doc.go
+
+// Package v1alpha1 contains API Schema definitions for the envoygateway.io API group.
+//
+// +groupName=config.envoygateway.io
+package v1alpha1
+```
+
+The initial `BootstrapConfig` API being proposed:
+```go
+// gateway/api/v1alpha1/bootstrapconfig_types.go
+
+// BootstrapConfigSpec defines the desired state of BootstrapConfig.
+type BootstrapConfigSpec struct {
+	// EnabledServices defines the services to enable. If unspecified,
+	// defaults to enabling all supported services.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=12
+	// +kubebuilder:validation:Enum=XdsServer;Provisioner
+	// +kubebuilder:default=XdsServer;Provisioner
+	EnabledServices []ServiceType `json:"enabledServices,omitempty"`
+
+	// XDSServer contains parameters of the xDS server. If unspecified,
+	// the xDS server will use default configuration settings.
+	//
+	// +optional
+	XDSServer *XDSServerConfig `json:"xdsServer,omitempty"`
```

Comment:
prefer if we used a higher level noun such as `ControlPlane` that had `xDSServer` as a field